posts/gsoc-2020-blog-4.md

```diff
@@ -0,0 +1,160 @@
+
+{{% thumbnail "/images/gsoc-2020/steady-state-eqs.png" align="center" %}}
+<!-- TEASER_END -->
+
+The system can now be solved with a nonlinear algebraic solver, which finds solution values for the state variables that will satisfy the system of equations. This means finding the zeroes to the functions defined on the right-hand-side of each equation. This is standard practice for any algebraic solution, and as such, nonlinear algebraic solvers typically require users to input a residual function that defines their specific system of equations.
+
+Cantera has a built-in damped Newton/time-stepping hybrid solver, defined in the `OneD` module. This solver was designed for use with multiple-domain one-dimensional problems—this means that it can solve problems with several linked one-dimensional *domains* (each domain can have a different number of state variables), where solutions can be found at *multiple spatial points* within each physical domain. I built `Cantera_NonLinSol`, a simplified interface to the 1D solver to solve *single* systems of nonlinear equations (see the implementation on [GitHub](https://github.com/paulblum/cantera/blob/0DSS/include/cantera/numerics/Cantera_NonLinSol.h)). This interface inputs problems to the Cantera 1D solver as single-domain, single-point systems. I used this interface within Cantera’s `ZeroD` module to solve the steady-state reactor equations and introduce basic steady-state solution capability.
+
+What follows is a general outline of how to use this solver, with specific implementation examples included from my work in `ZeroD`:
+
+#### *Include the `Cantera_NonLinSol.h` header in your code.*
+
+**ReactorNet.h**, line 12 (see this on [GitHub](https://github.com/paulblum/cantera/blob/ca36e253bd28c6d507eace5b6f1199cac64d8909/include/cantera/zeroD/ReactorNet.h#L12))
+
+```c++
+#include "cantera/numerics/Cantera_NonLinSol.h"
+```
+
+#### *Derive a class from `Cantera_NonLinSol` to inherit nonlinear algebraic solution capabilities and define your specific problem.*
+
+**ReactorNet.h**, line 24 (see this on [GitHub](https://github.com/paulblum/cantera/blob/ca36e253bd28c6d507eace5b6f1199cac64d8909/include/cantera/zeroD/ReactorNet.h#L24))
+
+```c++
+class ReactorNet : public FuncEval, public Cantera_NonLinSol
+```
+
+`ReactorNet` now exhibits *multiple inheritance*, since it was already a child of class `FuncEval`. The `FuncEval` parent class is used by `Integrator` during ODE time-integration, and its inheritance makes `ReactorNet` an *ODE right-hand-side evaluator*. (Confused? See my [last blog post](https://cantera.org/blog/gsoc-2020-blog-3)!)
+
+*Notice the difference in implementation here:* `Integrator` is an external solver object that uses user-defined functions from *another* class, `FuncEval`. This requires the inclusion of two headers, explicit initialization of an external object, and the passing of the `this` pointer:
```

Comment:
@paulblum this is _really_ well done--great work!  One question, here, this paragraph here took me a while to parse, as the comparison you are making is not obvious.  When you say "notice the difference in implementation here," what is it you are comparing?  Your approach versus what?  Is there a way to make that clearer?

Reply:
Thanks @decaluwe! Just committed an updated version of the post, let me know if it's any clearer.

Reply:
Looks great - thanks!